### PR TITLE
signal: manage shutdown requests with status codes

### LINKED ISSUE #5625: lnd should exit with no-success error code when automatically shut down due to bitcoind problems
### Background

While troubleshooting `tor` & `bitcoind` issues, I've restarted `bitcoind` two times in a row, and discovered (just accidentally, as I have `bitcoind` and `lnd` logs tailed in the same `tmux` split screen) that `lnd` is shutting itself down:

```
Aug 12 12:55:25 odroid-hc1 lnd[5606]: 2021-08-12 12:55:25.120 [ERR] NTFN: Unable to fetch block header: Post "http://127.0.0.1:8332": dial tcp 127.0.0.1:8332: connect: connection refused
Aug 12 12:55:25 odroid-hc1 lnd[5606]: 2021-08-12 12:55:25.304 [INF] CRTR: Pruning channel graph using block 0000000000000000000c7620d7807edfad1c475f204f5c17da4601e2a6e13945 (height=695396)
Aug 12 12:55:37 odroid-hc1 lnd[5606]: 2021-08-12 12:55:37.500 [INF] DISC: GossipSyncer(02ad6fb8d693dc1e4569bcedefadf5f72a931ae027dc0f0c544b34c1c6f3b9a02b): applying gossipFilter(start=0001-01-01 00:00:00 +0000 UTC, end=0001-01-01 00:00:00 +0000 UTC)
Aug 12 12:55:37 odroid-hc1 lnd[5606]: 2021-08-12 12:55:37.501 [INF] DISC: GossipSyncer(036d2ac71176151db04fdac839a0ddea9f3a584f6c23bb0b4ac72c323124ec506b): applying gossipFilter(start=2021-08-12 12:55:37.501421413 +0300 EEST m=+1873248.185437149, end=2157-09-18 19:23:52.501421413 +0300 EEST)
Aug 12 12:56:21 odroid-hc1 lnd[5606]: 2021-08-12 12:56:21.460 [INF] HLCK: Health check: chain backend, call: 2 failed with: -28: Verifying blocks..., backing off for: 2m0s
Aug 12 12:56:37 odroid-hc1 lnd[5606]: 2021-08-12 12:56:37.501 [INF] DISC: Broadcasting 49 new announcements in 5 sub batches
Aug 12 12:58:21 odroid-hc1 lnd[5606]: 2021-08-12 12:58:21.476 [CRT] SRVR: Health check: chain backend failed after 3 calls
Aug 12 12:58:21 odroid-hc1 lnd[5606]: 2021-08-12 12:58:21.476 [INF] SRVR: Sending request for shutdown
Aug 12 12:58:21 odroid-hc1 lnd[5606]: 2021-08-12 12:58:21.516 [INF] LTND: Received shutdown request.
Aug 12 12:58:21 odroid-hc1 lnd[5606]: 2021-08-12 12:58:21.517 [INF] LTND: Shutting down...
Aug 12 12:58:21 odroid-hc1 lnd[5606]: 2021-08-12 12:58:21.517 [INF] LTND: Gracefully shutting down.
...
Aug 12 12:59:58 odroid-hc1 lnd[5606]: 2021-08-12 12:59:58.385 [INF] LTND: Shutdown complete
Aug 12 12:59:58 odroid-hc1 systemd[1]: lnd.service: Succeeded.
```

`systemctl status` now is:
```
   Loaded: loaded (/etc/systemd/system/lnd.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Thu 2021-08-12 12:59:58 EEST; 9min ago
  Process: 5587 ExecStartPre=/usr/local/bin/lndcache.sh (code=exited, status=0/SUCCESS)
  Process: 5606 ExecStart=/home/lnd/lnd.go/bin/lnd --profile=9000 (code=exited, status=0/SUCCESS)
 Main PID: 5606 (code=exited, status=0/SUCCESS)
```
So it has `SUCCESS` status, meaning `systemd` will not restart `lnd` in case of this bail out, even if I have `Restart=on-failure`, as this was not reported as failure. This is risk for "silently" losing lightning functionality...

### Your environment

* `v0.13.1-beta`
* `Linux odroid-hc1 4.19.0-17-armmp-lpae #1 SMP Debian 4.19.194-3 (2021-07-18) armv7l GNU/Linux`
* `v0.21.1`

### Steps to reproduce

Keep restarting bitcoind until `lnd` shuts down with success result

### Expected behaviour

`lnd` process should exit with non-zero result.

### Actual behaviour

`lnd` exits with success result.

### PR DESCRIPTION
**Pull request by mohamedawnallah**
## Change Description

Replaces #5636
Closes #5625

## Steps to Test

0. Make sure you have configured chain backend health checks in the `lnd.conf` or you could pass it through `lncli`.
1. Run `bitcoind` daemon making sure it is functional.
2. Run `lnd` making sure it is functional and synced with the chain backend.
3. Fail `bitcoind` on purpose.
4. Exhaust all the attempts of checking the chain backend health check (the default is 3)
5. After that you will notice `lnd` shutting down gracefully
6. Get the previous status using this command `echo $?` the expected is `1` in our case previously it was `0`.

## Approach

Approach

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: signal/signal.go [hunk 1]
```diff
@@ -111,22 +111,28 @@ type Interceptor struct {
 	// close this channel.
 	quit chan struct{}
 
+	// exitCode represents a channel for receiving exit codes.
+	// It is used to communicate the exit status when the main
+	// interrupt handler exits.
+	exitCode chan int
```
Comment (ellemouton) [inline signal/signal.go:117]: why use a channel for this? channels for the rest make sense to me cause other sub-systems might wait on these but we would not wait on a specific exit code to be set. 

I think this can just be a mutex protected int and probably need a bool to indicate that we have in fact exited... ie, GetExitCode can return an int along with a bool. If the bool is false, it means we have not exited yet.

File: rpcserver.go [hunk 2]
```diff
@@ -630,7 +630,8 @@ var _ lnrpc.LightningServer = (*rpcServer)(nil)
 // dependencies are added, this will be an non-functioning RPC server only to
 // be used to register the LightningService with the gRPC server.
 func newRPCServer(cfg *Config, interceptorChain *rpcperms.InterceptorChain,
-	implCfg *ImplementationCfg, interceptor signal.Interceptor) *rpcServer {
+	implCfg *ImplementationCfg,
+	interceptor *signal.Interceptor) *rpcServer {
```
Comment (ellemouton) [inline rpcserver.go:634]: is the pointer change just for cleanliness? If so, should be in a separate commit (before the main commit)

File: docs/release-notes/release-notes-0.18.0.md [hunk 3]
```diff
@@ -274,6 +274,10 @@ bitcoin peers' feefilter values into account](https://github.com/lightningnetwor
   maintain a healthy connection to the network by checking the number of
   outbound peers if they are below 6.
 
+* [exitNoSuccessOnShutdown](https://github.com/lightningnetwork/lnd/pull/8659)
```
Comment (ellemouton) [inline docs/release-notes/release-notes-0.18.0.md:277]: `exitNoSuccessOnShutdown` is the branch name... doesnt mean much to the users reading the change log :)

File: signal/signal.go [hunk 4]
```diff
@@ -175,12 +182,27 @@ func (c *Interceptor) mainInterruptHandler() {
...
 			shutdown()
 
 		case <-c.shutdownRequestChannel:
 			log.Infof("Received shutdown request.")
+
+			// Set interceptor exit code status to 1 once we get a
+			// message from shutdownRequest Channel since it would
+			// exit with an error status after graceful shutdown
+			// when we close the quit channel.
+			c.setExitCode(1)
```
Comment (ellemouton) [inline signal/signal.go:201]: hmmm im not sure this is correct. 

`shutdownRequestChannel` could be closed if the `StopDaemon` rpc method is called. If LND shuts down smoothly after that, then the exit code should be 0, right?

Comment (ellemouton) [reply signal/signal.go:201]: perhaps instead `RequestShutdown` should have an error code param... so that call-sights can determine what the error code should be.... but then also, even if a call-sight indicates a `0` error code, the shutdown process itself should be able to downgrade that to a 1 ofc

**Review by ellemouton — CHANGES_REQUESTED** [review]
oops, clicked wrong button when submitting review :)

File: signal/signal.go (hunk 4 above)
Comment (ellemouton) [reply signal/signal.go:201]: > shutdownRequestChannel could be closed if the StopDaemon rpc method is called. If LND shuts down smoothly after that, then the exit code should be 0, right?

Ah - you can see this in action if you take a look at the failing itests. They all call `StopDaemon` to shutdown the various nodes but now exit code of each process is 1

File: docs/release-notes/release-notes-0.18.0.md (hunk 3 above)
Comment (mohamedawnallah) [reply docs/release-notes/release-notes-0.18.0.md:277]: Yeah, when I updated the release notes I didn't have text in mind to fit in the 80 chars limit (since URL takes much) so I went with the branch name :D.

Gonna update this!

**Comment by mohamedawnallah** [comment]
Side-Updates: Updated the approach diagram and the commit message to handle if the user called the `StopDaemon` RPC.

File: signal/signal.go [hunk 5]
```diff
@@ -175,12 +186,25 @@ func (c *Interceptor) mainInterruptHandler() {
...
+			c.setExitCode(0)
+
 			shutdown()
 
-		case <-c.shutdownRequestChannel:
+		case errorCode := <-c.shutdownRequestChannel:
 			log.Infof("Received shutdown request.")
+
+			// Set the interceptor exit code status based on the
+			// provided error code i.e We set exit code 1 for
```
Comment (Chinwendu20) [inline signal/signal.go:200]:
```suggestion
                       // provided error code i.e. we set exit code 1 for
```

File: signal/signal.go [hunk 6]
```diff
@@ -104,28 +105,37 @@ type Interceptor struct {
...
 	// quit is closed when instructing the main interrupt handler to exit.
 	// Note that to avoid losing notifications, only shutdown func may
 	// close this channel.
 	quit chan struct{}
 
+	// exitCode represents the exit code of the interceptor.
+	exitCode int
+
+	// exitCodeMutex protects the exitCode.
+	exitCodeMutex sync.Mutex
```
Comment (Chinwendu20) [inline signal/signal.go:119]: would exitCode be written to or read from multiple threads?

**Review by Chinwendu20 — COMMENTED** [review]
Learned a lot while reviewing this code and it is very thoughtful of you to include a diagram in your PR description. Huge kudos and thanks. Left some comments and question.

File: signal/signal.go [hunk 7]
```diff
@@ -227,3 +252,28 @@ func (c *Interceptor) RequestShutdown() {
 func (c *Interceptor) ShutdownChannel() <-chan struct{} {
 	return c.shutdownChannel
 }
+
+// GetExitCode returns the exit code of the interceptor, which represents
+// the exit status when the main interrupt handler exits. It blocks until
+// the exit code is available.
```
[the remark this answers is not in the document]
Comment (mohamedawnallah) [reply signal/signal.go:258]: Yes since we are waiting for a lock to be available here `c.exitCodeMutex.Lock()`.

File: signal/signal.go (hunk 6 above)
Comment (mohamedawnallah) [reply signal/signal.go:119]: Not sure what threads you are referring to here but multiple subsystems send a signal to `interruptChannel` or  `shutdownRequestChannel` we want to make sure there is no racing on this.

File: cmd/lnd/main.go [hunk 8]
```diff
@@ -41,4 +41,9 @@ func main() {
 		_, _ = fmt.Fprintln(os.Stderr, err)
 		os.Exit(1)
 	}
+
+	exitCode, exited := shutdownInterceptor.GetExitCode()
+	if exited {
```
[the remark this answers is not in the document]
Comment (mohamedawnallah) [reply cmd/lnd/main.go:46]: You're right I don't think we don't need it since the status code is returned. I maybe misunderstood @ellemouton comment:
> probably need a bool to indicate that we have in fact exited... ie, GetExitCode can return an int along with a bool. If the bool is false, it means we have not exited yet.

My view is that if we wait for a lock to be released and retrieve the status code (0/1), it implies the process has already exited so no need for additional flag.

File: signal/signal.go (hunk 6 above)
Comment (Chinwendu20) [reply signal/signal.go:119]: The comment on top of that referred to protecting the exitCode and not the channels which you are now referring to. Plus this mutex is not exported so other subsystems can't really call it. Additionally, I don't think it would result to a race condition if multiple channels send to a golang channel. What you might want to do is queue the signals for a more deterministic behaviour but that might be an overkill for this PR.

File: signal/signal.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Chinwendu20) [reply signal/signal.go:258]: IMO, I think writing that it is safe for concurrent use would be clearer.

Also, I do not think there is need for a behaviour to block in this function as this function is called when all other subsystems have stopped so the exit code should be readily available then.

File: signal/signal.go (hunk 6 above)
Comment (mohamedawnallah) [reply signal/signal.go:119]: > The comment on top of that referred to protecting the exitCode and not the channels which you are now referring to.

What I meant is When the subsystems send signals to the channels (interruptChannel and interruptChannel) this is where the code for setting the exit status code exists.

File: signal/signal.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mohamedawnallah) [reply signal/signal.go:258]: I agree with you, but since we protect `exitCode` with a mutex, it may be a good practice to wait for it. This is because we return the same `exitCode` in `GetExitCode`, although we're sure that the lock is not held ?

File: signal/signal.go (hunk 6 above)
Comment (Chinwendu20) [reply signal/signal.go:119]: > this is where the code for setting the exit status code exists so that's why we used mutex protected int to prevent racing on exit code.

Not an expert myself but from what you explained here and which I can see is what is happening in the code.
We do not have multiple subsystems setting the exit code. The `setExitCode` function is only called in one place the `mainInterruptHandler` goroutine. Once that is set, lnd and accompanying subsystems shutdown and so there would be no concurrency issue fetching that exitcode here: 
https://github.com/lightningnetwork/lnd/blob/f27a64ef48ac9fb5d33b15a0cd7c0990865e5914/cmd/lnd/main.go#L45

So we might not need the mutex?

File: signal/signal.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Chinwendu20) [reply signal/signal.go:258]: I am thinking that maybe the lock might not be needed? https://github.com/lightningnetwork/lnd/pull/8659#discussion_r1587545653

**Review by ellemouton — COMMENTED** [review]
If we take a step back, I think the main thing we need to happen here is we need `lnd.Main` to actually return an error. Ie, it should not be the case that `lnd.Main` returns a nil error but then we need to call `GetErrorCode` just in case. Especially since we are already waiting on a `signal` channel at the very end of the `lnd.Main` function. So instead I think we should listen on that & depending on what it returns, we should actually return an error from `lnd.Main`. 

Here is a _rough_ patch on top of your PR here showing an idea of how this can be simplified quite a bit:
[signal.patch](https://github.com/lightningnetwork/lnd/files/15334323/signal.patch)

My bad for leading you in the wrong direction on the last review! totally my fault - I should have thought it through a bit more. 

Let me know what you think of the idea in the patch 🙏

**Comment by guggero** [comment]
Closing as per but instructions. Feel free to re-open once comments are addressed.

**Comment by mohamedawnallah** [comment]
Thanks, @guggero! 🙏 I keep wanting to catch up on contributions to LND, but my current workload is heavy with my internship and schoolwork right now. 😅 Hope to be back soon! 🙏
